File: day-1/coin-change/bottom_up.cpp
// Problem : https://www.hackerrank.com/challenges/coin-change/problem
// BOTTOM_UP

#include <iostream>
#include <vector>

using namespace std;

long solve(int sum, vector<long> &denominations) {
    vector <long> memo(sum + 1, 0);
    memo[0] = 1;
    for(int i = 0; i < denominations.size(); i++) {
        for(int j = denominations[i]; j < sum + 1; j++) {
            memo[j] += memo[j - denominations[i]];
        }
    }
    return memo[sum];
}

int main() {
    int n, m;
    cin >> n >> m;

    vector<long> denominations(m);

    for (int i = 0; i < m; i++) {
        cin >> denominations[i];
    }

    cout << solve(n, denominations) << endl;
    return 0;
}

File: day-2/fibonacci-modified/bottom_up.py
# Problem : https://www.hackerrank.com/challenges/fibonacci-modified/problem
# NON_DP

def solve(n):
    if n < 0:
        return 0
    if memo[n] != -1:
        return memo[n]
    t = solve(n - 1)
    memo[n] = t * t + solve(n - 2)
    return memo[n]

a, b, n = [int(x) for x in input().strip().split(' ')]

memo = [-1 for _ in range(n + 1)]
memo[0] = a
memo[1] = b
memo[2] = a + b * b

print(solve(n - 1))